### ISSUE TITLE
Specifying type when using pattern matching

### ISSUE THREAD

**Issue by ytimenkov** [microsoft/pyright#4663]
Python 3.11.1, VS Code python extension v2023.2.0

I'm using pyright in strict mode and for the following function

```python
def test(d1: Any, d2: Any):
    match (d1, d2):
        case dict(a), dict(b):
            print(f"Two dicts: {a}, {b}")
        case _:
            print("Something else")
```

it complains with:
```
(variable) a: dict[Unknown, Unknown]
Type of "a" is partially unknown
  Type of "a" is "dict[Unknown, Unknown]"

Pylance[reportUnknownVariableType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnknownVariableType)
```

I think specifying generic type is now possible with pattern matching, but somehow hinting type checker would be nice.

I'm probably missing something, but specifying non-generic dict is not possible either (or at least in strict mode).

Or is strict mode is not suitable for such cases at all? (I think I tried basic mode, but it was too relaxed...).

**Comment by erictraut** [comment]
Pyright is working as designed here. This isn't a bug.

There is no way for a type checker to know the specialized type of a `dict` when you're narrowing it from `Any`. The best it can do is narrow to `dict[Unknown, Unknown]`. So this is expected behavior. It's effectively the same as an `isinstance(x, dict)` conditional check.

The approach I recommend is to not use `Any` as the input parameter type. If you instead use a union of potential types (e.g. `dict[str, str] | str | list[str]`), then there will be no `Unknown` generated during the type narrowing process.

**Comment by ytimenkov** [comment]
Got it! 🙇‍♂️

I think the first repro I tried with a stronger typed value didn't raise any warning, so I will give it a try.

But I find it annoying that pyright complains on the expression in the `case` statement for a `dict` and there is no way around it except disabling the warning. In the following code it is possible to narrow down the type via assert or cast.
